File: Payload_Type/freyja/mythic/agent_functions/cmd_executor.py
from mythic_payloadtype_container.MythicCommandBase import *
import json
from mythic_payloadtype_container.MythicRPC import *


class CmdArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = []

    async def parse_arguments(self):
        pass


class CmdCommand(CommandBase):
    cmd = "cmd_executor"
    needs_admin = False
    help_cmd = "cmd_executor {command}"
    description = "Execute a shell command using 'cmd /C'"
    version = 1
    author = "@antman1p"
    attackmapping = ["T1059", "T1059.003"]
    argument_class = CmdArguments
    attributes = CommandAttributes(
        suggested_command=True,
        supported_os=[SupportedOS.Windows]
    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        resp = await MythicRPC().execute("create_artifact", task_id=task.id,
            artifact="cmd /C {}".format(task.args.command_line),
            artifact_type="Process Create",
        )
        resp = await MythicRPC().execute("create_artifact", task_id=task.id,
            artifact="{}".format(task.args.command_line),
            artifact_type="Process Create",
        )
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/freyja/freyja/agentfunctions/cmd_executor.go
package agentfunctions

import (
	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
	"github.com/MythicMeta/MythicContainer/logging"
	"github.com/MythicMeta/MythicContainer/mythicrpc"
)

var cmd_executor = agentstructs.Command{
	Name:                "cmd_executor",
	Author:              "@antman1p",
	Description:         "Execute a shell command using 'cmd /C'",
	MitreAttackMappings: []string{"T1059"},
	CommandAttributes: agentstructs.CommandAttribute{
		SupportedOS: []string{agentstructs.SUPPORTED_OS_WINDOWS},
	},
	TaskFunctionCreateTasking: cmd_executorCreateTasking,
	Version:                   1,
}

func init() {
	agentstructs.AllPayloadData.Get("freyja").AddCommand(cmd_executor)
}

func cmd_executorCreateTasking(taskData *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
	response := agentstructs.PTTaskCreateTaskingMessageResponse{
		Success: true,
		TaskID:  taskData.Task.ID,
	}
	if _, err := mythicrpc.SendMythicRPCArtifactCreate(mythicrpc.MythicRPCArtifactCreateMessage{
		BaseArtifactType: "ProcessCreate",
		ArtifactMessage:  "cmd /C " + taskData.Args.GetCommandLine(),
		TaskID:           taskData.Task.ID,
	}); err != nil {
		logging.LogError(err, "Failed to send mythicrpc artifact create")
	}
	return response
}
